services/server/server.go

```diff
@@ -540,41 +537,6 @@ func (s *Server) ChangeSettings(ctx context.Context, req *serverpb.ChangeSetting
 	}, nil
 }
 
-func (s *Server) validateAlertManagerRules(ctx context.Context, rules string) error {
-	tempFile, err := ioutil.TempFile("", "temp_rules_*.yml")
-	if err != nil {
-		return errors.WithStack(err)
-	}
-	tempFile.Close()                 //nolint:errcheck
-	defer os.Remove(tempFile.Name()) //nolint:errcheck
-
-	if err = ioutil.WriteFile(tempFile.Name(), []byte(rules), 0644); err != nil {
-		return errors.WithStack(err)
-	}
```

Comment:
`TempFile` returns already prepared for writing file, it's not obvious why we need to close it and open it again with `WriteFile`. And those file permissions `0644` are never used, as the file already exists. 🤷‍♂️

Reply:
> it's not obvious why we need to close it and open it again with WriteFile

Exactly because WriteFile opens and closes this file. Otherwise, we have to be very careful about closing and removing a file – that is a common source of FS-dependent problems and leaked file descriptors ([example](https://github.com/percona/pmm-agent/blob/e39464ccbc770e23fb453b58deac68bba25141de/agents/mysql/slowlog/slowlog.go#L237-L267)). I gave a suggestion above how to improve it, but I think that old code is even more straight-forward.